daemon/cdi_other.go

```diff
@@ -0,0 +1,9 @@
+//go:build !linux
+// +build !linux
+
+package daemon
+
+// registerCDIDriver is a no-op for non-Linux platforms.
```

Comment:
There are no fundamental limitations that require linux device for CDI, but the OCI spec injection code implemented in the cdi package is currently linux-specific. With that said, I don't know what the timeline would be for extending this for other platforms (e.g. Windows), but I suppose reworking the code here at that time would make more sense than overcomplicating it now.

Reply:
> the OCI spec injection code implemented in the cdi package is currently linux-specific

That's exactly the sort of technical limitation I was looking for. Please document that information so the next person to look at this code knows why the limitation is in place [and when it can be lifted.](https://thoughtbot.com/blog/chestertons-fence)

Reply:
I realise https://github.com/container-orchestrated-devices/container-device-interface/issues/28 exists, but AFAICT [the general CDI injection code](https://github.com/container-orchestrated-devices/container-device-interface/blob/fca21545c4761414626d1632888497c211a9f799/pkg/cdi/container-edits.go) is _not_ Linux-specific, [just the support for copying attributes from host-device nodes](https://github.com/container-orchestrated-devices/container-device-interface/blob/fca21545c4761414626d1632888497c211a9f799/pkg/cdi/container-edits_windows.go), i.e. the current test, which is env-var-setting-only, _should_ be workable on Windows as far as CDI goes.

That said, the Docker DeviceRequests implementation setup is currently Linux-only, so until that's taken care of (perhaps by the refactoring to use dependency injection as mooted elsewhere in this PR) an integration test won't be implementable.

(I also just noticed a bug in the CDI code I linked, as it will fail even if the device node specification is already complete, which blocks, e.g., LCOW support, although that's not really an issue for Docker until and if we connect up containerd for Windows and the uvm-based LCOWv2)

Reply:
No, the general code is not linux-specific, but as you pointed out, the function to retrieve information from the host device is (and is implemented as a no-op on non-linux platforms). One thing that is not clear on first reading is that the cal to [`specgen.RemoveDevice()` and `specgen.AddDevice()`](https://github.com/container-orchestrated-devices/container-device-interface/blob/fca21545c4761414626d1632888497c211a9f799/pkg/cdi/container-edits.go#L104-L105) assume linux devices, so these would also need to replaced with equivalent calls `specgen.AddWindowsDevices`.

I would also expect that the CDI spec would need some extension to cleanly handle Windows devices.

> (I also just noticed a bug in the CDI code I linked, as it will fail even if the device node specification is already complete, which blocks, e.g., LCOW support, although that's not really an issue for Docker until and if we connect up containerd for Windows and the uvm-based LCOWv2)

Could you update https://github.com/container-orchestrated-devices/container-device-interface/issues/131 with more contex. (I will try to come up with a quick reproducer later today if I get a chance).

Reply:
> One thing that is not clear on first reading is that the cal to [`specgen.RemoveDevice()` and `specgen.AddDevice()`](https://github.com/container-orchestrated-devices/container-device-interface/blob/fca21545c4761414626d1632888497c211a9f799/pkg/cdi/container-edits.go#L104-L105) assume linux devices, so these would also need to replaced with equivalent calls `specgen.AddWindowsDevices`.

That's a different problem, https://github.com/container-orchestrated-devices/container-device-interface/issues/28 is tracking that.

I'm talking about Windows-hosted builds specifying _Linux_ device nodes, e.g. for LCOW cases, or I guess kata-containers or similar VM-managed systems.

I've added a comment to https://github.com/container-orchestrated-devices/container-device-interface/issues/131 to try and clarify, and am happy to continue discussion there if I'm still unclear, since none of the cases where this matters apply to Docker today.